### ISSUE TITLE
Segmentation fault at vlan.h

### ISSUE THREAD

**Issue by ymansour** [snort3/snort3#60]
Snort 3 Beta faults with error "Unable to find a Codec with data link type 113". In Snort 2, this solved with the option "--enable-non-ether-decoders". However, Snort 3 does not appear to have this option.

```
[Thread 0x7ffff2001700 (LWP 16068) exited]
-- [2] ./pcaps/fuzz-2014-04-11-8332.pcap
++ [2] ./pcaps/fuzz-2014-04-20-7766.pcap
[New Thread 0x7ffff2001700 (LWP 16076)]
ERROR: Unable to find a Codec with data link type 113

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff3003700 (LWP 16075)]
0x000000000055ba7a in snort::vlan::VlanTagHdr::vid (this=0x0) at /root/sources/snort3/src/protocols/vlan.h:41
41	    { return ntohs(vth_pri_cfi_vlan) & 0x0FFF; }
#0  0x000000000055ba7a in snort::vlan::VlanTagHdr::vid (this=0x0) at /root/sources/snort3/src/protocols/vlan.h:41
#1  0x000000000055ea67 in FlowControl::set_key (this=0x7fffef2950f0, key=0x7ffff2fcc3c0, p=0x7fffef5281f0) at /root/sources/snort3/src/flow/flow_control.cc:177
#2  0x000000000055f418 in FlowControl::process (this=0x7fffef2950f0, type=TCP, p=0x7fffef5281f0) at /root/sources/snort3/src/flow/flow_control.cc:365
#3  0x0000000000645b51 in StreamBase::eval (this=0x205c910, p=0x7fffef5281f0) at /root/sources/snort3/src/stream/base/stream_base.cc:244
#4  0x00000000005d9c8f in execute (p=0x7fffef5281f0, prep=0x5e4b570, num=1) at /root/sources/snort3/src/managers/inspector_manager.cc:912
#5  0x00000000005d9ea0 in snort::InspectorManager::execute (p=0x7fffef5281f0) at /root/sources/snort3/src/managers/inspector_manager.cc:963
#6  0x000000000052f497 in snort::DetectionEngine::inspect (p=0x7fffef5281f0) at /root/sources/snort3/src/detection/detection_engine.cc:403
#7  0x00000000005c76d9 in snort::Snort::process_packet (p=0x7fffef5281f0, pkthdr=0x7ffff2fcc5e0, pkt=0x2a05e01c "", is_frag=false) at /root/sources/snort3/src/main/snort.cc:892
#8  0x00000000005c7a3e in snort::Snort::packet_callback (pkthdr=0x7ffff2fcc5e0, pkt=0x2a05e01c "") at /root/sources/snort3/src/main/snort.cc:996
#9  0x000000000079f231 in pcap_process_loop (user=user@entry=0x1e4be00 "", pkth=pkth@entry=0x7ffff2fcc690, data=0x2a05e01c "") at daq_pcap.c:376
#10 0x00007ffff759b0d1 in pcap_offline_read (p=0x2a1a62c0, cnt=-1, callback=0x79f1a0 <pcap_process_loop>, user=0x1e4be00 "") at ./savefile.c:506
#11 0x000000000079f35b in pcap_daq_acquire (handle=0x1e4be00, cnt=0, callback=<optimized out>, metaback=<optimized out>, user=<optimized out>) at daq_pcap.c:394
#12 0x00000000005fdcd9 in snort::SFDAQInstance::acquire (this=0x2a16ca30, max=0, callback=0x5c791c <snort::Snort::packet_callback(void*, _daq_pkthdr const*, unsigned char const*)>) at /root/sources/snort3/src/packet_io/sfdaq.cc:570
#13 0x00000000005b157a in Analyzer::analyze (this=0x5e4a130) at /root/sources/snort3/src/main/analyzer.cc:161
#14 0x00000000005b136e in Analyzer::operator() (this=0x5e4a130, ps=0xcb5b1b0, run_num=1429) at /root/sources/snort3/src/main/analyzer.cc:99
#15 0x000000000052954e in std::__invoke<Analyzer<Swapper*, unsigned short> > (__f=...) at /usr/include/c++/4.8.2/functional:234
#16 0x00000000005294ed in std::reference_wrapper<Analyzer>::operator()<Swapper*, unsigned short>(Swapper*&&, unsigned short&&) const (this=0x7fffec112950) at /usr/include/c++/4.8.2/functional:467
#17 0x0000000000529451 in std::_Bind_simple<std::reference_wrapper<Analyzer> (Swapper*, unsigned short)>::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (this=0x7fffec112940) at /usr/include/c++/4.8.2/functional:1732
#18 0x0000000000529309 in std::_Bind_simple<std::reference_wrapper<Analyzer> (Swapper*, unsigned short)>::operator()() (this=0x7fffec112940) at /usr/include/c++/4.8.2/functional:1720
#19 0x00000000005292a2 in std::thread::_Impl<std::_Bind_simple<std::reference_wrapper<Analyzer> (Swapper*, unsigned short)> >::_M_run() (this=0x7fffec112928) at /usr/include/c++/4.8.2/thread:115
#20 0x00007ffff528b070 in ?? () from /lib64/libstdc++.so.6
#21 0x00007ffff6b42e25 in start_thread () from /lib64/libpthread.so.0
#22 0x00007ffff49efbad in clone () from /lib64/libc.so.6
```

**Comment by traceflight** [comment]
This is because the pcap file is captured from **any** device. The format is [Linux cooked-mode capture (SLL)](https://wiki.wireshark.org/SLL) which is not supported by snort. Your should capture network traffic from physical device such as **eth0**.

**Comment by Xiche** [comment]
That segfault was due to a different PCAP where it actually started processing packets (as can be seen from the backtrace).  Did you just take the default Snort config and run it against all of the Wireshark fuzzed PCAPs (https://www.wireshark.org/download/automated/captures/)?

**Comment by ymansour** [comment]
@Xiche,

Correct, it was a different pcap since I can see the same messages without a segfault. Yes, the pcaps were from the Wireshark fuzzed collection. Snort config is slightly modified.

snort.lua
```
require('snort_config')

-- this depends on SNORT_LUA_PATH
-- where to find other config files
conf_dir = os.getenv('SNORT_LUA_PATH')

if ( not conf_dir ) then
    conf_dir = '.'
end

---------------------------------------------------------------------------
-- 2. configure defaults
---------------------------------------------------------------------------

-- HOME_NET and EXTERNAL_NET must be set now
-- setup the network addresses you are protecting
-- HOME_NET = 'any'
HOME_NET = [[ 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 ]]

-- set up the external network addresses.
-- (leave as "any" in most situations)
EXTERNAL_NET = 'any'

dofile(conf_dir .. '/snort_defaults.lua')
dofile(conf_dir .. '/file_magic.lua')

---------------------------------------------------------------------------
-- 3. configure inspection
---------------------------------------------------------------------------

-- mod = { } uses internal defaults
-- you can see them with snort --help-module mod

-- mod = default_mod uses external defaults
-- you can see them in snort_defaults.lua

-- the following are quite capable with defaults:

stream = { }
stream_ip = { }
stream_icmp = { }
stream_tcp = { }
stream_udp = { }
stream_user = { }
stream_file = { }

arp_spoof = { }
back_orifice = { }
dnp3 = { }
dns = { }
http_inspect = { }
imap = { }
modbus = { }
normalizer = { }
pop = { }
rpc_decode = { }
sip = { }
ssh = { }
ssl = { }
telnet = { }

dce_smb = { }
dce_tcp = { }
dce_udp = { }
dce_http_proxy = { }
dce_http_server = { }

-- see snort_defaults.lua for default_*
gtp_inspect = default_gtp
port_scan = default_med_port_scan
smtp = default_smtp

ftp_server = default_ftp_server
ftp_client = { }
ftp_data = { }

-- see file_magic.lua for file id rules
-- file_id = { file_rules = file_magic }

file_id =
{
    file_rules = file_magic,
    file_policy =
    {
        { when = { file_type_id = 22 }, use = { verdict = 'log', enable_file_signature = true } },
        { when = { sha256 = "E65ECCC561CACE1860638CD0BC745E59058F16349F7455E215BDDF3233355007" }, use = { verdict = 'log' } }
    }
}


-- the following require additional configuration to be fully effective:

appid =
{
    -- appid requires this to use appids in rules
    app_detector_dir = APPID_PATH,
    log_stats = true
}


reputation =
{
    -- configure one or both of these, then uncomment reputation
    -- scan_local = true,
    blacklist = BLACK_LIST_PATH .. '/ip-blacklist',
    whitelist = WHITE_LIST_PATH .. '/ip-whitelist'
}

---------------------------------------------------------------------------
-- 4. configure bindings
---------------------------------------------------------------------------

wizard = default_wizard

binder =
{
    -- port bindings required for protocols without wizard support
    { when = { proto = 'udp', ports = '53' },  use = { type = 'dns' } },
    { when = { proto = 'tcp', ports = '111' }, use = { type = 'rpc_decode' } },
    { when = { proto = 'tcp', ports = '502' }, use = { type = 'modbus' } },
    { when = { proto = 'tcp', ports = '2123 2152 3386' }, use = { type = 'gtp' } },

    { when = { proto = 'tcp', service = 'dcerpc' }, use = { type = 'dce_tcp' } },
    { when = { proto = 'udp', service = 'dcerpc' }, use = { type = 'dce_udp' } },

    { when = { service = 'netbios-ssn' },      use = { type = 'dce_smb' } },
    { when = { service = 'dce_http_server' },  use = { type = 'dce_http_server' } },
    { when = { service = 'dce_http_proxy' },   use = { type = 'dce_http_proxy' } },

    { when = { service = 'dnp3' },             use = { type = 'dnp3' } },
    { when = { service = 'dns' },              use = { type = 'dns' } },
    { when = { service = 'ftp' },              use = { type = 'ftp_server' } },
    { when = { service = 'ftp-data' },         use = { type = 'ftp_data' } },
    { when = { service = 'gtp' },              use = { type = 'gtp_inspect' } },
    { when = { service = 'imap' },             use = { type = 'imap' } },
    { when = { service = 'http' },             use = { type = 'http_inspect' } },
    { when = { service = 'modbus' },           use = { type = 'modbus' } },
    { when = { service = 'pop3' },             use = { type = 'pop' } },
    { when = { service = 'ssh' },              use = { type = 'ssh' } },
    { when = { service = 'sip' },              use = { type = 'sip' } },
    { when = { service = 'smtp' },             use = { type = 'smtp' } },
    { when = { service = 'ssl' },              use = { type = 'ssl' } },
    { when = { service = 'sunrpc' },           use = { type = 'rpc_decode' } },
    { when = { service = 'telnet' },           use = { type = 'telnet' } },

    { use = { type = 'wizard' } }
}

---------------------------------------------------------------------------
-- 5. configure performance
---------------------------------------------------------------------------

-- use latency to monitor / enforce packet and rule thresholds
latency =
{
    packet = { max_time = 1500 },
    rule = { max_time = 200 },
}

-- use these to capture perf data for analysis and tuning
--profiler = { }
--perf_monitor = { }

---------------------------------------------------------------------------
-- 6. configure detection
---------------------------------------------------------------------------

references = default_references
classifications = default_classifications

ips =
{
    mode = tap,

    -- use this to enable decoder and inspector alerts
    enable_builtin_rules = true,

    -- use include for rules files; be sure to set your path
    -- note that rules files can include other rules files
    --include = 'snort3_community.rules'

    -- The following include syntax is only valid for BUILD_243 (13-FEB-2018) and later
    -- RULE_PATH is typically set in snort_defaults.lua
    rules = [[

        include $RULE_PATH/snort3-app-detect.rules
        include $RULE_PATH/snort3-browser-chrome.rules
        include $RULE_PATH/snort3-browser-firefox.rules
        include $RULE_PATH/snort3-browser-ie.rules
        include $RULE_PATH/snort3-browser-other.rules
        include $RULE_PATH/snort3-browser-plugins.rules
        include $RULE_PATH/snort3-browser-webkit.rules
        include $RULE_PATH/snort3-content-replace.rules
        include $RULE_PATH/snort3-exploit-kit.rules
        include $RULE_PATH/snort3-file-executable.rules
        include $RULE_PATH/snort3-file-flash.rules
        include $RULE_PATH/snort3-file-identify.rules
        include $RULE_PATH/snort3-file-image.rules
        include $RULE_PATH/snort3-file-java.rules
        include $RULE_PATH/snort3-file-multimedia.rules
        include $RULE_PATH/snort3-file-office.rules
        include $RULE_PATH/snort3-file-other.rules
        include $RULE_PATH/snort3-file-pdf.rules
        include $RULE_PATH/snort3-indicator-compromise.rules
        include $RULE_PATH/snort3-indicator-obfuscation.rules
        include $RULE_PATH/snort3-indicator-scan.rules
        include $RULE_PATH/snort3-indicator-shellcode.rules
        include $RULE_PATH/snort3-malware-backdoor.rules
        include $RULE_PATH/snort3-malware-cnc.rules
        include $RULE_PATH/snort3-malware-other.rules
        include $RULE_PATH/snort3-malware-tools.rules
        include $RULE_PATH/snort3-netbios.rules
        include $RULE_PATH/snort3-os-linux.rules
        include $RULE_PATH/snort3-os-mobile.rules
        include $RULE_PATH/snort3-os-other.rules
        include $RULE_PATH/snort3-os-solaris.rules
        include $RULE_PATH/snort3-os-windows.rules
        include $RULE_PATH/snort3-policy-multimedia.rules
        include $RULE_PATH/snort3-policy-other.rules
        include $RULE_PATH/snort3-policy-social.rules
        include $RULE_PATH/snort3-policy-spam.rules
        include $RULE_PATH/snort3-protocol-dns.rules
        include $RULE_PATH/snort3-protocol-finger.rules
        include $RULE_PATH/snort3-protocol-ftp.rules
        include $RULE_PATH/snort3-protocol-icmp.rules
        include $RULE_PATH/snort3-protocol-imap.rules
        include $RULE_PATH/snort3-protocol-nntp.rules
        include $RULE_PATH/snort3-protocol-other.rules
        include $RULE_PATH/snort3-protocol-pop.rules
        include $RULE_PATH/snort3-protocol-rpc.rules
        include $RULE_PATH/snort3-protocol-scada.rules
        include $RULE_PATH/snort3-protocol-services.rules
        include $RULE_PATH/snort3-protocol-snmp.rules
        include $RULE_PATH/snort3-protocol-telnet.rules
        include $RULE_PATH/snort3-protocol-tftp.rules
        include $RULE_PATH/snort3-protocol-voip.rules
        include $RULE_PATH/snort3-pua-adware.rules
        include $RULE_PATH/snort3-pua-other.rules
        include $RULE_PATH/snort3-pua-p2p.rules
        include $RULE_PATH/snort3-pua-toolbars.rules
        include $RULE_PATH/snort3-server-apache.rules
        include $RULE_PATH/snort3-server-iis.rules
        include $RULE_PATH/snort3-server-mail.rules
        include $RULE_PATH/snort3-server-mssql.rules
        include $RULE_PATH/snort3-server-mysql.rules
        include $RULE_PATH/snort3-server-oracle.rules
        include $RULE_PATH/snort3-server-other.rules
        include $RULE_PATH/snort3-server-samba.rules
        include $RULE_PATH/snort3-server-webapp.rules
        include $RULE_PATH/snort3-sql.rules
        include $RULE_PATH/snort3-x11.rules

    ]]

}

alert_fast =
{
    file = true
}

file_log =
{
    log_pkt_time = true,
    log_sys_time = false,
}

data_log = 
{
    key = 'http_request_header_event',
    limit = 100
}
```

**Comment by Xiche** [comment]
Ok, thanks a lot for reporting these.  I've run with the default config + ODP against the full set of Wireshark fuzzed PCAPs and reproduced all three of your reported crashes locally.  We'll be working to fix them soon.

**Comment by Xiche** [comment]
This finally got fixed in b04040b157a009de869435399a3a9ae940d97f44.  Thank you for reporting this!
